### PR TITLE
#168781836 Users can view or create their accommodation facilities

### PR DESCRIPTION
**Pull request by lemoissonM**
#### What does this PR do?
creates accommodation endpoints 
#### Description of Task to be completed?
- adds accommodations tests
- adds accommodation controllers: **create accommodation and get accommodations**
- add accommodations middlewares
- add accommodations routes
- add accommodations documentation

#### How should this be manually tested?
Run npm i to install relevant packages

**add new environments in .env file**

**POST api/v1/accommodation to add an accommodation** 
Data to send in the body*
``` description, address, availableSpace, cost, services, amenities```
**GET api/v1/accommodation to get accommodations**

For these two routes, 
**Put the token in the authorization header**
 **to get the token :**
run npm run db-seeder:dev
then signin using email : alexismajyambere@gmail.com && password : 12345678UUi@
`` use the token provided in the response body``

#### Any background context you want to provide?

Screen Shot 2019-10-14 at 16 54 53

#### What are the relevant pivotal tracker stories?
https://www.pivotaltracker.com/story/show/168781836

#### Screenshots (if appropriate)
Screen Shot 2019-10-16 at 09 11 18
Screen Shot 2019-10-16 at 09 11 25

#### Questions:

### CODE REVIEW

File: src/database/migrations/20191001134215-create-accommodation.js [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+      }
+    });
+  },
+  down: (queryInterface, Sequelize) => {
+    return queryInterface.dropTable('Accommodations',  {
+      force: true,
+      cascade: true,
+  });
+  }
+};
```
Comment (swaibat) [inline src/database/migrations/20191001134215-create-accommodation.js]: @lemoissonM good job but there is a need for a new line here

File: src/database/migrations/20191001134218-create-accommodation-image.js [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+      },},
+      imageurl: { type: Sequelize.STRING, allowNull: false, },
+      createdAt: { allowNull: false, type: Sequelize.DATE, defaultValue: Sequelize.NOW },
+      updatedAt: { allowNull: false, type: Sequelize.DATE},
+    });
+  },
+  down: (queryInterface, Sequelize) => {
+    return queryInterface.dropTable('AccommodationImages');
+  }
+};
```
Comment (swaibat) [inline src/database/migrations/20191001134218-create-accommodation-image.js]: @lemoissonM good job but there is a need for a new line here

File: src/database/models/Accommodations.js [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+            foreignKey: 'userid',
+            targetKey: 'id',
+          });
+    Accommodation.hasMany(models.AccommodationImages, {
+      foreignKey: 'accommodationid',
+      targetKey: 'id',
+    });
+    };
+    return Accommodation;
+  };
```
Comment (swaibat) [inline src/database/models/Accommodations.js]: @lemoissonM good job but there is a need for a new line here

File: src/database/models/accommodationImage.js [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+      imageurl: DataTypes.STRING,
+    }, { freezeTableName: true });
+    AccommodationIm.associate = function(models) {
+         AccommodationIm.belongsTo(models.Accommodations, {
+          foreignKey: 'accommodationid',  
+          targetKey: 'id',
+          });        
+    };
+    return AccommodationIm;
+  };
```
Comment (swaibat) [inline src/database/models/accommodationImage.js]: @lemoissonM good job but there is a need for a new line here

**Review by urbanishimwe — CHANGES_REQUESTED** [review]
Big work. however, try to delete these `.DS_Store`

**Review by swaibat — CHANGES_REQUESTED** [review]

File: src/database/migrations/20191001134213-create-user.js [hunk 5]
```diff
@@ -7,12 +7,16 @@ module.exports = {
...
     });
   },
   down: (queryInterface, Sequelize) => {
-    return queryInterface.dropTable('Users');
+    return queryInterface.dropTable('Users', {
+      force: true,
+      cascade: true,
+  });
   }
 };
```
Comment (swaibat) [inline src/database/migrations/20191001134213-create-user.js]: new line after

**Review by swaibat — CHANGES_REQUESTED** [review]

**Comment by lemoissonM** [comment]
> Big work. however, try to delete these `.DS_Store`

I've put them in gitignore but, they are still coming, and they are invisible in vscode

File: src/middlewares/accommodationMiddleware.js [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    const { files } = req;
+    let imageArray = [];
+    if (!files) {
+      return sendResult(res, 400, 'upload at least one image');
+    }
+    const { images } = files;
+    if (Array.isArray(images)) { imageArray = images; } else imageArray.push(images);
+    // eslint-disable-next-line arrow-parens
+    const wrongFile = imageArray.find(file => !file.mimetype || !file.mimetype.startsWith('image'));
+    if (wrongFile) return sendResult(res, 400, 'uploaded files should be images');
```
Comment (swaibat) [inline src/middlewares/accommodationMiddleware.js:48]: great job @lemoissonM  but I don't see the validation of supported image formats

**Review by swaibat — CHANGES_REQUESTED** [review]

File: src/middlewares/accommodationMiddleware.js (hunk 6 above)
Comment (lemoissonM) [reply src/middlewares/accommodationMiddleware.js:48]: All images are supported!

Comment (lemoissonM) [reply src/middlewares/accommodationMiddleware.js:48]: other files are of course prohibitted

File: src/database/models/Accommodations.js [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+
```
Comment (swaibat) [inline src/database/models/Accommodations.js]: @lemoissonM great job but there is still need for a new line here

File: src/middlewares/accommodationMiddleware.js (hunk 6 above)
Comment (swaibat) [reply src/middlewares/accommodationMiddleware.js:48]: alright

File: src/routes/accommodationRoute.js [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+ *           items:
+ *             $ref: '#/definitions/Accommodation'
+ *       401:
+ *         description: You are not authorized
+ */
+const fUpload = fileUpload({
+  useTempFiles: true,
+  tempFileDir: path.join(__dirname, '../temp'),
+});
+app.post('/', fUpload, userMidd.checkToken, accMidd.checkUserSupplier, accMidd.validateAccommodation, accMidd.checkForImages, accommodationCont.addAccommodation);
```
Comment (swaibat) [inline src/routes/accommodationRoute.js]: @lemoissonM consider destructuring or creating a new line the route exceeds the required length

**Review by swaibat — CHANGES_REQUESTED** [review]

File: src/database/models/Accommodations.js (hunk 7 above)
Comment (lemoissonM) [reply src/database/models/Accommodations.js]: there is a new line, don't you see?

File: src/controllers/accommodationController.js [hunk 9]
```diff
@@ -0,0 +1,60 @@
+import sendResult from '../utils/sendResult';
+import db from '../database/models/index';
+import cloudinary from '../config/clound-config';
+
+let success = 0;
+let failure = 0;
+const checkIfAllUploaded = (total, res, data) => {
+  if (success + failure === total) return sendResult(res, 201, 'created successfuly', data);
+};
```
Comment (nakiwuge) [inline src/controllers/accommodationController.js:10]: Add a new line here

File: src/database/models/Accommodations.js (hunk 7 above)
Comment (swaibat) [reply src/database/models/Accommodations.js]: always first remove trailing spaces before adding a new line

File: src/controllers/accommodationController.js [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+const uploadImages = (req, res, data) => {
+  const tot = req.imageArray.length;
```
Comment (nakiwuge) [inline src/controllers/accommodationController.js]: Please give this constant a better name

File: src/controllers/accommodationController.js [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+  data.images = [];
+  req.imageArray.forEach((element) => {
+    cloudinary.uploader.upload(element.tempFilePath, async (result, error) => {
+      if (error) { failure += 1; checkIfAllUploaded(tot, res, data); }
```
Comment (nakiwuge) [inline src/controllers/accommodationController.js]: Add a new line here

**Review by swaibat — COMMENTED** [review]
@lemoissonM  great job but also and the environments for Cloudinary in `.sample.env`

File: src/controllers/accommodationController.js [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+      const imageRes = await db.AccommodationImages.create({
+        imageurl: result.url,
+        accommodationid: data.id,
+      });
+      data.images.push(imageRes.imageurl);
+      success += 1;
+      checkIfAllUploaded(tot, res, data);
+    });
+  });
+};
```
Comment (nakiwuge) [inline src/controllers/accommodationController.js:28]: Add a new line here

File: src/controllers/accommodationController.js [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+      userid: userId,
+      availableSpace,
+    });
+    const accommodation = response.get({ plain: true });
+    uploadImages(req, res, accommodation);
+  },
+
+  async getAccommodation(req, res) {
+    const { role, userId } = req.userData;
+    if (role === 'ASupp') {
```
Comment (nakiwuge) [inline src/controllers/accommodationController.js]: The role name should be `supplier or host`

File: src/controllers/accommodationController.js [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+      const accommodations = await db.Accommodations.findAll({
+        where: { userid: userId },
+        include: [{ model: db.AccommodationImages, attributes: { exclude: ['id', 'accommodationid', 'createdAt', 'updatedAt'] } }],
+      });
+      return sendResult(res, 200, 'my accommodations', accommodations);
+    }
```
Comment (nakiwuge) [inline src/controllers/accommodationController.js:59]: Add new line

File: src/controllers/accommodationController.js [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+    const accommodations = await db.Accommodations.findAll();
+    return sendResult(res, 200, 'all accommodations', accommodations);
+  },
+
+
```
Comment (nakiwuge) [inline src/controllers/accommodationController.js]: Remove this line

File: src/middlewares/accommodationMiddleware.js [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+    const { userData } = req;
+    if (userData.role === 'ASupp' || userData.role === 'TAdmin') {
+      return next();
+    }
+    return sendResult(res, 401, 'You are not authorized');
+  },
+
+  checkForImages(req, res, next) {
+    const { files } = req;
+    let imageArray = [];
```
Comment (nakiwuge) [inline src/middlewares/accommodationMiddleware.js:36]: Add a new line here

File: src/middlewares/accommodationMiddleware.js [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+    if (!files) {
+      return sendResult(res, 400, 'upload at least one image');
+    }
```
Comment (nakiwuge) [inline src/middlewares/accommodationMiddleware.js:40]: Add a new line here

File: src/middlewares/accommodationMiddleware.js [hunk 18]
```diff
@@ -0,0 +1,50 @@
...
+    const { images } = files;
+    if (Array.isArray(images)) { imageArray = images; } else imageArray.push(images);
```
Comment (nakiwuge) [inline src/middlewares/accommodationMiddleware.js:43]: Add a new line here

File: src/middlewares/userMiddlware.js [hunk 19]
```diff
@@ -38,6 +39,15 @@ const User = {
       ? sendResult(res, 400, password.error)
       : next();
   },
+  checkToken(req, res, next) {
+    const token = helper.getToken(req);
+    const data = helper.verifyToken(token);
```
Comment (nakiwuge) [inline src/middlewares/userMiddlware.js:45]: Add a new line here

File: src/routes/accommodationRoute.js [hunk 20]
```diff
@@ -0,0 +1,86 @@
...
+ *       401:
+ *         description: You are not authorized
+ */
+const fUpload = fileUpload({
+  useTempFiles: true,
+  tempFileDir: path.join(__dirname, '../temp'),
+});
+const { checkUserSupplier, validateAccommodation, checkForImages } = accMidd;
+const { checkToken } = userMidd;
+const { addAccommodation, getAccommodation } = accommodationCont;
```
Comment (nakiwuge) [inline src/routes/accommodationRoute.js:83]: Add a new line here

File: src/tests/accommodation.spec.js [hunk 21]
```diff
@@ -0,0 +1,184 @@
...
+import app from '../index';
+import mockLogin from './mockData/loginMockData';
+import helper from '../utils/helper';
+
+chai.use(chaiHttp);
+const { expect } = chai;
+let Utoken = '';
+const img = fs.readFileSync(path.join(__dirname, 'mockData/img.jpeg'));
+const img2 = fs.readFileSync(path.join(__dirname, 'mockData/img2.jpeg'));
+const wrongFile = fs.readFileSync(path.join(__dirname, 'mockData/email.js'));
```
Comment (nakiwuge) [inline src/tests/accommodation.spec.js:17]: Add a new line here

**Review by nakiwuge — CHANGES_REQUESTED** [review]
Good job @lemoissonM . I have left comments on this pr. Kindly check them out

**Review by urbanishimwe — APPROVED** [review]

**Review by alexismajya — APPROVED** [review]

**Review by swaibat — APPROVED** [review]

**Review by nakiwuge — APPROVED** [review]

**Pull request merged**